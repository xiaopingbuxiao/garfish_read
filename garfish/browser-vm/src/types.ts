import { Sandbox } from './sandbox';
import type { LoaderOptions } from '@garfish/loader';

export type FakeWindow = Window & Record<PropertyKey, any>;
export type Module = (sandbox: Sandbox) => OverridesData | void;

export interface OverridesData {
  recover?: () => void;
  prepare?: () => void;
  created?: (context: Sandbox['global']) => void;
  override?: Record<PropertyKey, any>;
}

export interface ReplaceGlobalVariables {
  recoverList: Array<OverridesData['recover']>;
  prepareList: Array<OverridesData['prepare']>;
  createdList: Array<OverridesData['created']>;
  overrideList: Record<PropertyKey, any>;
}

export interface SandboxOptions {
  namespace: string;
  baseUrl?: string;
  fixBaseUrl?: boolean;
  disableWith?: boolean;
  strictIsolation?: boolean;
  modules?: Array<Module>;
  addSourceList?: (sourceInfo: Array<{ tagName: string; url: string }> | { tagName: string; url: string }) => void;
  loaderOptions?: LoaderOptions;
  styleScopeId?: () => string;
  el?: () => Element | ShadowRoot | null;
  protectVariable?: () => Array<PropertyKey>;
  insulationVariable?: () => Array<PropertyKey>;
}
