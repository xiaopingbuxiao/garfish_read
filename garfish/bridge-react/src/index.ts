export { reactBridge } from './reactBridge';
export type { PropsInfo } from './types';
