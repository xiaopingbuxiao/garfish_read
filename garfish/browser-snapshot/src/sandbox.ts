import { PatchEvent } from './patchers/event';
import { PatchStyle } from './patchers/style';
import { PatchHistory } from './patchers/history';
import { PatchInterval } from './patchers/interval';
import { PatchGlobalVal } from './patchers/variable';
import { PatchWebpackJsonp } from './patchers/webpackjsonp';

export class Sandbox {
  public type = 'snapshot';
  public isRunning: Boolean = false;
  private patchList: Array<
    PatchGlobalVal | PatchStyle | PatchInterval | PatchEvent | PatchWebpackJsonp
  > = [];

  constructor(
    public name: string,
    public protectVariable: Array<PropertyKey> = [],
    public targetToProtect: Window | Object = typeof window !== 'undefined'
      ? window
      : globalThis,
    private isInBrowser: Boolean = typeof window === 'undefined' ? false : true,
  ) {
    this.name = name;
    this.isInBrowser = isInBrowser;
    this.patchList.push(new PatchGlobalVal(targetToProtect, protectVariable));

    // 执行顺序是，全局环境变量先激活，全局环境变量后销毁
    if (this.isInBrowser) {
      this.patchList = [
        ...this.patchList,
        new PatchStyle(),
        new PatchEvent(),
        new PatchHistory(),
        new PatchInterval(),
        new PatchWebpackJsonp(),
      ];
    }
  }

  //  1.触发生命周期钩子，willActivate（将要激活）
  //  2.将当前组的其他沙盒disable，并触发switch生命周期
  //  3.将当前window对象属性进行缓存
  //  4.获取style节点，进行缓存
  //  5.恢复沙盒运行期间产生的副作用
  public activate() {
    if (this.isRunning) return;
    this.patchList.forEach((patch) => {
      patch.activate();
    });
    this.isRunning = true;
  }

  // 1.恢复沙盒启动期间变量变更的变量，记录变更记录
  // 2.恢复沙盒启动期间删除的变量，记录变更记录
  public deactivate(clearEffects: boolean = true) {
    if (!this.isRunning) return; // 最后销毁全局变量守护
    [...this.patchList].reverse().forEach((patch) => {
      patch.deactivate(clearEffects);
    });
    this.isRunning = false;
  }
}
