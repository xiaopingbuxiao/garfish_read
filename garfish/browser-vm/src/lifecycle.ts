import { SyncHook, SyncWaterfallHook, PluginSystem } from '@garfish/hooks';
import type { interfaces } from '@garfish/core';
import type { FakeWindow } from './types';

export interface DocumentGetterData {
  value: any;
  propName: PropertyKey;
  proxyDocument: Document;
  rootNode?: null | Element | ShadowRoot;
  customValue?: any;
}

export function sandboxLifecycle() {
  return new PluginSystem({
    closed: new SyncHook<[], void>(),
    stared: new SyncHook<[FakeWindow?], void>(),
    appendNode: new SyncHook<[Element, Element, Element, string], void>(),
    documentGetter: new SyncWaterfallHook<DocumentGetterData>('documentGetter'),
    beforeClearEffect: new SyncHook<[], void>(),
    afterClearEffect: new SyncHook<[], void>(),
    beforeInvoke: new SyncHook<
      [
        { code: string },
        string?,
        Record<string, any>?,
        interfaces.ExecScriptOptions?,
      ],
      void
    >(),
    afterInvoke: new SyncHook<
      [
        { code: string },
        string?,
        Record<string, any>?,
        interfaces.ExecScriptOptions?,
      ],
      void
    >(),
    invokeError: new SyncHook<
      [Error, string?, Record<string, any>?, interfaces.ExecScriptOptions?],
      void
    >(),
  });
}
