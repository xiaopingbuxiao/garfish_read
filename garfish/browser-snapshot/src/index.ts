import { Sandbox } from './sandbox';
import { interfaces } from '@garfish/core';
import './globalExtensions';

export { Sandbox as default } from './sandbox';

export interface SandboxConfig {
  snapshot?: boolean;
  disableWith?: boolean;
  strictIsolation?: boolean;
}

interface BrowserConfig {
  open?: boolean;
  protectVariable?: PropertyKey[];
}

export function GarfishBrowserSnapshot(op?: BrowserConfig) {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    const options = {
      openBrowser: false,
      version: __VERSION__,
      name: 'browser-snapshot',

      afterLoad(appInfo, appInstance) {
        const config: BrowserConfig = op || { open: true };
        const sandboxConfig = appInfo.sandbox || Garfish?.options?.sandbox;
        if (
          sandboxConfig === false ||
          sandboxConfig.open === false ||
          sandboxConfig?.snapshot === false
        ) {
          config.open = false;
        }
        if (sandboxConfig) {
          config.protectVariable = [
            ...(Garfish?.options.protectVariable || []),
            ...(appInfo.protectVariable || []),
          ];
        }
        options.openBrowser = !!config.open;
        if (!config.open) return;
        if (appInstance) {
          // existing
          if (appInstance.snapshotSandbox) return;
          const sandbox = new Sandbox(appInfo.name, config.protectVariable);
          appInstance.snapshotSandbox = sandbox;
        }
      },

      beforeMount(appInfo, appInstance) {
        // existing
        if (!appInstance.snapshotSandbox) return;
        appInstance.snapshotSandbox.activate();
      },

      afterUnmount(appInfo, appInstance) {
        if (!appInstance.snapshotSandbox) return;
        appInstance.snapshotSandbox.deactivate();
      },
    };
    return options;
  };
}
