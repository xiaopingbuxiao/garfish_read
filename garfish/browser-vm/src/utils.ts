import { hasOwn, makeMap, nextTick } from '@garfish/utils';
import { Sandbox } from './sandbox';
import { FakeWindow } from './types';
import {
  __proxyNode__,
  __sandboxMap__,
  __elementSandboxTag__,
} from './symbolTypes';

// https://tc39.es/ecma262/#sec-function-properties-of-the-global-object
const esGlobalMethods =
  // Function properties of the global object // Function properties of the global object
  (
    'eval,isFinite,isNaN,parseFloat,parseInt,' +
    // URL handling functions
    'decodeURI,decodeURIComponent,encodeURI,encodeURIComponent,' +
    // Constructor properties of the global object
    'Array,ArrayBuffer,BigInt,BigInt64Array,BigUint64Array,Boolean,DataView,Date,Error,EvalError,' +
    'FinalizationRegistry,Float32Array,Float64Array,Function,Int8Array,Int16Array,Int32Array,Map,Number,' +
    'Object,Promise,Proxy,RangeError,ReferenceError,RegExp,Set,SharedArrayBuffer,String,Symbol,SyntaxError,' +
    'TypeError,Uint8Array,Uint8ClampedArray,Uint16Array,Uint32Array,URIError,WeakMap,WeakRef,WeakSet,' +
    // Other Properties of the Global Object
    'Atomics,JSON,Math,Reflect,'
  ).split(',');

const nativeCodeMethods = 'hasOwnProperty,'.split(',');

export const isEsGlobalMethods = makeMap(esGlobalMethods);
export const isNativeCodeMethods = makeMap(nativeCodeMethods);

// Need to optimize, avoid from the with
// Can't filter document, eval keywords, such as document in handling parentNode useful
export const optimizeMethods = [...esGlobalMethods].filter((v) => v !== 'eval');

// The sandbox may be used alone, to ensure that the `sandboxMap` is globally unique,
// because we will only rewrite `appendChild` once
let sandboxList: Map<number, Sandbox> = new Map();
if (!(window as FakeWindow)[__sandboxMap__]) {
  (window as FakeWindow)[__sandboxMap__] = sandboxList;
} else {
  sandboxList = (window as FakeWindow)[__sandboxMap__];
}

export const sandboxMap = {
  sandboxMap: sandboxList,

  get(element: Element): Sandbox | undefined {
    if (!element) return;
    const sandboxId = element[__elementSandboxTag__];
    if (typeof sandboxId !== 'number') return;
    return this.sandboxMap.get(sandboxId);
  },

  setElementTag(element: Element, sandbox: Sandbox) {
    if (!element) return;
    element[__elementSandboxTag__] = sandbox.id;
  },

  set(sandbox: Sandbox) {
    if (this.sandboxMap.get(sandbox.id)) return;
    this.sandboxMap.set(sandbox.id, sandbox);
  },

  del(sandbox: Sandbox) {
    this.sandboxMap.delete(sandbox.id);
  },
};

export function handlerParams(args: IArguments | Array<any>) {
  args = Array.isArray(args) ? args : Array.from(args);
  return args.map((v) => {
    return v && v[__proxyNode__] ? v[__proxyNode__] : v;
  });
}

// Container node, because it changes all the time, take it as you use it
export function rootElm(sandbox: Sandbox) {
  const container = sandbox && sandbox.options.el;
  return container && container();
}

export function isInIframe() {
  return window?.parent?.__GARFISH__ !== window?.__GARFISH__;
}

// Copy "window" and "document"
export function createFakeObject(
  target: Record<PropertyKey, any>,
  filter?: (key: PropertyKey) => boolean,
  isWritable?: (key: PropertyKey) => boolean,
) {
  const fakeObject = {};
  const propertyMap = {};
  const storageBox = Object.create(null); // Store changed value
  const propertyNames = Object.getOwnPropertyNames(target);
  const def = (p: string) => {
    const descriptor = Object.getOwnPropertyDescriptor(target, p);

    if (descriptor?.configurable) {
      const hasGetter = hasOwn(descriptor, 'get');
      const hasSetter = hasOwn(descriptor, 'set');
      const canWritable = typeof isWritable === 'function' && isWritable(p);

      if (hasGetter) {
        // prettier-ignore
        descriptor.get = () => hasOwn(storageBox, p)
          ? storageBox[p]
          : target[p];
      }
      if (hasSetter) {
        descriptor.set = (val) => {
          storageBox[p] = val;
          return true;
        };
      }
      if (canWritable) {
        if (descriptor.writable === false) {
          descriptor.writable = true;
        } else if (hasGetter) {
          descriptor.set = (val) => {
            storageBox[p] = val;
            return true;
          };
        }
      }
      Object.defineProperty(fakeObject, p, Object.freeze(descriptor));
    }
  };
  propertyNames.forEach((p) => {
    propertyMap[p] = true;
    typeof filter === 'function' ? !filter(p) && def(p) : def(p);
  });
  // "prop" maybe in prototype chain
  for (const prop in target) {
    !propertyMap[prop] && def(prop);
  }
  return fakeObject as any;
}

let setting = true;
export function microTaskHtmlProxyDocument(proxyDocument) {
  // The HTML parent node into agent for the document
  // In micro tasks replace primary node
  const html = document.children[0];
  if (html && html.parentNode !== proxyDocument) {
    Object.defineProperty(html, 'parentNode', {
      value: proxyDocument,
      configurable: true,
    });

    if (setting) {
      setting = false;
      // Do not use micro tasks, Element will appear in the task placed in nextTick after node
      nextTick(() => {
        setting = true;
        Object.defineProperty(html, 'parentNode', {
          value: document,
          configurable: true,
        });
      });
    }
  }
}

export function isStyledComponentsLike(element: HTMLStyleElement) {
  // A styled-components liked element has no textContent but keep the rules in its sheet.cssRules.
  return (
    element instanceof HTMLStyleElement &&
    !element.textContent &&
    element.sheet?.cssRules.length
  );
}
