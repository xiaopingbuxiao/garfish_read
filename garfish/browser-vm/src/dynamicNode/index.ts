import { makeMap, safeWrapper, warn } from '@garfish/utils';
import { StyleManager } from '@garfish/loader';
import { __domWrapper__ } from '../symbolTypes';
import { injectHandlerParams } from './processParams';
import { DynamicNodeProcessor, rawElementMethods } from './processor';
import { isInIframe, sandboxMap, isStyledComponentsLike } from '../utils';
import { SandboxOptions } from '../types';

const mountElementMethods = [
  'append',
  'appendChild',
  'insertBefore',
  'insertAdjacentElement',
];
const removeChildElementMethods = ['removeChild'];

const ignoreElementTimingTags = makeMap([
  'STYLE',
  'SCRIPTS',
  'LINK',
  'META',
  'TITLE',
]);

function injector(current: Function, methodName: string) {
  return function (this: Element) {
    // prettier-ignore
    const el = methodName === 'insertAdjacentElement'
      ? arguments[1]
      : arguments[0];
    const sandbox = sandboxMap.get(el);
    const originProcess = () => current.apply(this, arguments);

    if (sandbox) {
      if (el && this?.tagName?.toLowerCase() === 'style') {
        const manager = new StyleManager(el.textContent);
        const { baseUrl, namespace, styleScopeId } = sandbox.options;
        manager.correctPath(baseUrl);
        manager.setScope({
          appName: namespace,
          rootElId: styleScopeId!(),
        });
        el.textContent = manager.transformCode(manager.styleCode);
        return originProcess();
      } else {
        const processor = new DynamicNodeProcessor(el, sandbox, methodName);
        return processor.append(this, arguments, originProcess);
      }
    }

    // custom performance Element Timing API
    // https://web.dev/custom-metrics/#element-timing-api
    safeWrapper(() => {
      if (ignoreElementTimingTags(el.tagName)) return;
      if (
        el?.setAttribute &&
        typeof el?.setAttribute === 'function' &&
        !el?.getAttribute('elementtiming')
      ) {
        el?.setAttribute(
          'elementtiming',
          sandbox
            ? `${(sandbox as any).options.namespace}-element-timing`
            : 'element-timing',
        );
      }
    });

    if (sandbox) {
      const processor = new DynamicNodeProcessor(el, sandbox, methodName);
      return processor.append(this, arguments, originProcess);
    } else {
      return originProcess();
    }
  };
}

function injectorRemoveChild(current: Function, methodName: string) {
  return function (this: Element) {
    const el = arguments[0];
    const sandbox = el && sandboxMap.get(el);
    const originProcess = () => {
      // Sandbox may have applied sub dom side effects to delete
      // by removeChild deleted by the tag determine whether have been removed
      return current.apply(this, arguments);
    };

    if (sandbox) {
      const processor = new DynamicNodeProcessor(el, sandbox, methodName);
      return processor.removeChild(this, originProcess);
    }
    return originProcess();
  };
}

// Handle `ownerDocument` to prevent elements created by `ownerDocument.createElement` from escaping
function handleOwnerDocument() {
  Object.defineProperty(window.Element.prototype, 'ownerDocument', {
    get() {
      const sandbox = this && sandboxMap.get(this);
      const realValue = Reflect.get(
        window.Node.prototype,
        'ownerDocument',
        this,
      );
      return sandbox ? sandbox.global.document : realValue;
    },
    set() {
      __DEV__ && warn('"ownerDocument" is a read-only attribute.');
    },
  });
}

export function makeElInjector(sandboxConfig: SandboxOptions) {
  if ((makeElInjector as any).hasInject) return;
  (makeElInjector as any).hasInject = true;

  if (typeof window.Element === 'function') {
    // iframe can read html container this can't point to proxyDocument has Illegal invocation error
    if (sandboxConfig.fixBaseUrl) safeWrapper(()=> handleOwnerDocument());
    const rewrite = (
      methods: Array<string>,
      builder: typeof injector | typeof injectorRemoveChild,
    ) => {
      for (const name of methods) {
        const fn = window.Element.prototype[name];
        if (typeof fn !== 'function' || fn[__domWrapper__]) {
          continue;
        }
        rawElementMethods[name] = fn;
        const wrapper = builder(fn, name);
        wrapper[__domWrapper__] = true;
        window.Element.prototype[name] = wrapper;
      }
    };
    rewrite(mountElementMethods, injector);
    rewrite(removeChildElementMethods, injectorRemoveChild);
  }

  injectHandlerParams();
}

export function recordStyledComponentCSSRules(
  dynamicStyleSheetElementSet: Set<HTMLStyleElement>,
  styledComponentCSSRulesMap: WeakMap<HTMLStyleElement, CSSRuleList>,
) {
  dynamicStyleSheetElementSet.forEach((styleElement) => {
    if (isStyledComponentsLike(styleElement) && styleElement.sheet) {
      styledComponentCSSRulesMap.set(styleElement, styleElement.sheet.cssRules);
    }
  });
}

export function rebuildCSSRules(
  dynamicStyleSheetElementSet: Set<HTMLStyleElement>,
  styledComponentCSSRulesMap: WeakMap<HTMLStyleElement, CSSRuleList>,
) {
  dynamicStyleSheetElementSet.forEach((styleElement) => {
    const cssRules = styledComponentCSSRulesMap.get(styleElement);
    if (cssRules && (isStyledComponentsLike(styleElement) || cssRules.length)) {
      for (let i = 0; i < cssRules.length; i++) {
        const cssRule = cssRules[i];
        // re-insert rules for styled-components element
        styleElement.sheet?.insertRule(
          cssRule.cssText,
          styleElement.sheet?.cssRules.length,
        );
      }
    }
  });
}
