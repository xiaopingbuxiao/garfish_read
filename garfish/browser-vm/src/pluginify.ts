import { interfaces } from '@garfish/core';
import { warn, isPlainObject } from '@garfish/utils';
import { Module } from './types';
import { Sandbox } from './sandbox';
import { recordStyledComponentCSSRules, rebuildCSSRules } from './dynamicNode';

declare module '@garfish/core' {
  export default interface Garfish {
    setGlobalValue(key: string, value?: any): void;
    getGlobalObject: () => Window & typeof globalThis;
    clearEscapeEffect: (key: string, value?: any) => void;
  }

  export namespace interfaces {
    export interface SandboxConfig {
      modules?: Array<Module> | Record<string, Module>;
    }

    export interface Config {
      protectVariable?: PropertyKey[];
      insulationVariable?: PropertyKey[];
    }

    export interface AppInfo {
      protectVariable?: PropertyKey[];
      insulationVariable?: PropertyKey[];
    }

    export interface App {
      vmSandbox?: Sandbox;
    }
  }
}

const specialExternalVariables = [
  'onerror',
  'webpackjsonp',
  '__REACT_ERROR_OVERLAY_GLOBAL_HOOK__',
  __DEV__ ? 'webpackHotUpdate' : '',
];

function compatibleOldModule(
  modules: Array<Module> | Record<string, Module>,
): Array<Module> {
  if (!Array.isArray(modules)) {
    __DEV__ && warn('"vm sandbox" modules should be an array');
    const list: Array<Module> = [];
    for (const key in modules) {
      list.push(modules[key]);
    }
    modules = list;
  }
  return modules;
}

function rewriteAppAndSandbox(
  Garfish: interfaces.Garfish,
  app: interfaces.App,
  sandbox: Sandbox,
) {
  const originExecScript = sandbox.execScript;
  // Rewrite sandbox attributes
  sandbox.loader = Garfish.loader;
  sandbox.execScript = (code, env, url, options) => {
    const evalHooksArgs = [app.appInfo, code, env, url, options] as const;
    app.hooks.lifecycle.beforeEval.emit(...evalHooksArgs);
    try {
      const res = originExecScript.call(
        sandbox,
        code,
        {
          // For application of environment variables
          ...env,
          ...app.getExecScriptEnv(options?.noEntry),
        },
        url,
        options,
      );
      app.hooks.lifecycle.afterEval.emit(...evalHooksArgs);
      return res;
    } catch (err) {
      app.hooks.lifecycle.errorExecCode.emit(err, ...evalHooksArgs);
      throw err;
    }
  };

  // Rewrite app attributes
  app.vmSandbox = sandbox;
  app.global = sandbox.global;
  app.strictIsolation = sandbox.options.strictIsolation ?? false;
  app.runCode = function () {
    return originExecScript.apply(sandbox, arguments);
  };
  if (app.entryManager.DOMApis && sandbox.global) {
    app.entryManager.DOMApis.document = sandbox.global.document;
  }
}

function createOptions(Garfish: interfaces.Garfish) {
  const canSupport = Sandbox.canSupport();

  const options: interfaces.Plugin = {
    name: 'browser-vm',
    version: __VERSION__,

    afterLoad(appInfo, appInstance) {
      if (
        !canSupport ||
        !appInstance ||
        appInstance?.vmSandbox ||
        appInfo.sandbox === false || // Ensure that old versions compatible
        (appInfo.sandbox && appInfo.sandbox.open === false) ||
        (appInfo.sandbox && appInfo.sandbox.snapshot)
      ) {
        if (appInstance?.vmSandbox) {
          appInstance.global = appInstance.vmSandbox.global;
        }
        return;
      }

      rewriteAppAndSandbox(
        Garfish,
        appInstance,
        new Sandbox({
          namespace: appInfo.name,
          addSourceList: appInstance.addSourceList.bind(appInstance),
          baseUrl: appInstance.entryManager.url,
          modules: compatibleOldModule(appInfo.sandbox?.modules || []),
          fixBaseUrl: Boolean(appInfo.sandbox?.fixBaseUrl),
          disableWith: Boolean(appInfo.sandbox?.disableWith),
          strictIsolation: Boolean(appInfo.sandbox?.strictIsolation),

          el: () => appInstance.htmlNode,
          styleScopeId: () => appInstance.appContainer.id,
          protectVariable: () => appInfo.protectVariable || [],
          insulationVariable: () => {
            return [
              ...specialExternalVariables,
              ...(appInfo.insulationVariable || []),
            ].filter(Boolean);
          },
        }),
      );
    },

    beforeUnmount(appInfo, appInstance) {
      if (appInstance.vmSandbox) {
        recordStyledComponentCSSRules(
          appInstance.vmSandbox.dynamicStyleSheetElementSet,
          appInstance.vmSandbox.styledComponentCSSRulesMap,
        );
      }
    },

    // If the app is uninstalled, the sandbox needs to clear all effects and then reset
    afterUnmount(appInfo, appInstance, isCacheMode) {
      // The caching pattern to retain the same context
      if (appInstance.vmSandbox && !isCacheMode) {
        appInstance.vmSandbox.reset();
      }
    },

    afterMount(appInfo, appInstance) {
      if (appInstance.vmSandbox) {
        rebuildCSSRules(
          appInstance.vmSandbox.dynamicStyleSheetElementSet,
          appInstance.vmSandbox.styledComponentCSSRulesMap,
        );
        appInstance.vmSandbox.execScript(`
          if (typeof window.onload === 'function') {
            window.onload.call(window);
          }
        `);
      }
    },
  };
  return options;
}

// Export Garfish plugin
export function GarfishBrowserVm() {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    Garfish.getGlobalObject = function () {
      return Sandbox.getNativeWindow();
    };

    Garfish.setGlobalValue = function (key, value) {
      return (this.getGlobalObject()[key] = value);
    };

    Garfish.clearEscapeEffect = function (key, value) {
      const global = this.getGlobalObject();
      if (key in global) {
        global[key] = value;
      }
    };
    return createOptions(Garfish);
  };
}
