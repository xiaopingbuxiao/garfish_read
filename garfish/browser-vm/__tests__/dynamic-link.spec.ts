import { mockStaticServer } from '@garfish/test-suite';
import { __MockHead__ } from '@garfish/utils';
import { Sandbox } from '../src/sandbox';

declare global {
  interface window {
    execOrder: Array<string>;
    fetch: any;
  }
}

describe('Sandbox: dynamic link', () => {
  const withSuffix = '/resources/links/suffix.css';
  const withoutSuffix = '/resources/links/no-suffix';
  const withoutSuffixAndContentType = '/resources/links/no-content-type-suffix';
  const styleType = 'text/css';
  const jsonType = 'application/json';
  mockStaticServer({
    baseDir: __dirname,
    customerHeaders: {
      [withSuffix]: {
        'Content-Type': jsonType,
      },
      [withoutSuffix]: {
        'Content-Type': styleType,
      },
      [withoutSuffixAndContentType]: {
        'Content-Type': jsonType,
      },
    },
  });

  let sandbox: Sandbox;

  const go = (code: string) => {
    return `
      const sandbox = __debug_sandbox__;
      const Sandbox = sandbox.constructor;
      const nativeWindow = Sandbox.getNativeWindow();
      ${code}
    `;
  };

  beforeEach(() => {
    document.body.innerHTML = `<div id="root">123<div ${__MockHead__}></div></div>`;
    sandbox = new Sandbox({
      namespace: 'app',
      el: () => document.querySelector('#root'),
      modules: [
        () => ({
          recover() {},
          override: { go, jest, expect },
        }),
      ],
    });
    (global as any).fetch = fetch;
  });

  it('without ref', (done) => {
    sandbox.execScript(
      go(`
        const dynamicLink = document.createElement('link');
        dynamicLink.href = "${withSuffix}";
        document.head.appendChild(dynamicLink);
        expect(document.body).toMatchSnapshot();
        jestDone();
      `),
      { jestDone: done },
    );
  });

  it('with suffix', (done) => {
    sandbox.execScript(
      go(`
        const dynamicLink = document.createElement('link');
        dynamicLink.href = "${withSuffix}";
        dynamicLink.rel = 'stylesheet';
        dynamicLink.onload = function () {
          expect(document.body).toMatchSnapshot();
          jestDone();
        }
        document.head.appendChild(dynamicLink);
      `),
      { jestDone: done },
    );
  });

  it('with content-type', (done) => {
    sandbox.execScript(
      go(`
        const dynamicLink = document.createElement('link');
        dynamicLink.href = "${withoutSuffix}";
        dynamicLink.rel = 'stylesheet';
        dynamicLink.onload = function () {
          expect(document.body).toMatchSnapshot();
          jestDone();
        }
        document.head.appendChild(dynamicLink);
      `),
      { jestDone: done },
    );
  });

  it('without suffix and content-type', (done) => {
    sandbox.execScript(
      go(`
        const dynamicLink = document.createElement('link');
        dynamicLink.href = "${withoutSuffixAndContentType}";
        dynamicLink.rel = 'stylesheet';
        dynamicLink.onload = function () {
          expect(document.body).toMatchSnapshot();
          jestDone();
        }
        document.head.appendChild(dynamicLink);
      `),
      { jestDone: done },
    );
  });
});
