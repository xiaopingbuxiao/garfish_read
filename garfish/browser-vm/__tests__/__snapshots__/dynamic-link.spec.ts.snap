// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sandbox: dynamic link with content-type 1`] = `
<body>
  <div
    id="root"
  >
    123
    <div
      __garfishmockhead__=""
    >
      <!--&lt;link src="http://localhost/resources/links/no-suffix" execute by garfish(dynamic)&gt;&lt;/link&gt;-->
      <style
        type="text/css"
      >
        /* no-suffix test case */
#root{
  color: red;
  background: yellow;
}

      </style>
    </div>
  </div>
</body>
`;

exports[`Sandbox: dynamic link with suffix 1`] = `
<body>
  <div
    id="root"
  >
    123
    <div
      __garfishmockhead__=""
    >
      <!--&lt;link src="http://localhost/resources/links/suffix.css" execute by garfish(dynamic)&gt;&lt;/link&gt;-->
      <style
        type="text/css"
      >
        /* suffix test case */
#root{
  color: red;
  background: blue;
}

      </style>
    </div>
  </div>
</body>
`;

exports[`Sandbox: dynamic link without ref 1`] = `
<body>
  <div
    id="root"
  >
    123
    <div
      __garfishmockhead__=""
    >
      <link
        href="/resources/links/suffix.css"
      />
    </div>
  </div>
</body>
`;

exports[`Sandbox: dynamic link without suffix and content-type 1`] = `
<body>
  <div
    id="root"
  >
    123
    <div
      __garfishmockhead__=""
    >
      <!--&lt;link src="http://localhost/resources/links/no-content-type-suffix" execute by garfish(dynamic)&gt;&lt;/link&gt;-->
    </div>
  </div>
</body>
`;
